compare_skeleton.py

```diff
@@ -0,0 +1,240 @@
+from scipy.sparse import csr_matrix
+import pickle
+import networkx as nx
+import matplotlib.pyplot as plt
+from zss import simple_distance
+from nltk.metrics.distance import edit_distance
+import argparse
+import numpy as np
+
+
+class Node:
+    def __init__(self, label):
+        self.label = label
+        self.children = []
+
+def build_zss_tree(nx_tree, root_label):
+    root = Node(root_label)
+    _build_tree_helper(nx_tree, root, root_label, set())
+    return root
+
+def _build_tree_helper(nx_tree, current_node, current_label, visited):
+    visited.add(current_label)
+    children = []
+    for neighbor in nx_tree.neighbors(current_label):
+        if neighbor not in visited:
+            child_node = Node(neighbor)
+            children.append(child_node)
+            _build_tree_helper(nx_tree, child_node, neighbor, visited)
+    current_node.children = sorted(children, key=lambda x: x.label) # sorts node!
+
+
+def all_paths_to_leaves(tree, root):
+    paths = []
+    for leaf in [x for x in tree.nodes() if tree.degree(x) == 1 and x != root]:
+        paths.append(nx.shortest_path(tree, root, leaf))
+    return paths
+
+def min_edit_distance(paths1, paths2):
+    min_edit_dic= {}
+    for path1 in paths1:
+        for path2 in paths2:
+            
+            distance = edit_distance(path1, path2)/(len(path1)-1)
+            if str(path1) not in min_edit_dic:
+                min_edit_dic[str(path1)] = distance
+            else:
+                min_edit_dic[str(path1)]  = min(min_edit_dic[str(path1)], distance)
+    return min_edit_dic
+
+def reindex_tree_by_attribute(tree, attribute='syn_id'):
+    '''
+    This function reindex the tree with syn_id for easier comparision
+    '''
+    new_tree = type(tree)()
+    for node, data in tree.nodes(data=True):
+        if attribute in data:
+            new_tree.add_node(data[attribute])
+        else:
+            raise ValueError(f"Node {node} does not have the attribute '{attribute}'.")
+    for u, v in tree.edges():
+        new_u = tree.nodes[u][attribute]
+        new_v = tree.nodes[v][attribute]
+        new_tree.add_edge(new_u, new_v)
+    return new_tree
+
+def draw_trees_side_by_side(tree1, tree2, title1, title2, path = None):
+    '''
+    This function draws the two trees side by side
+    '''
+    fig, axs = plt.subplots(1, 2, figsize=(12, 6))
+    pos = nx.spring_layout(tree1)  
+    nx.draw(tree1, pos, with_labels=True, ax=axs[0], node_size=200, node_color='skyblue',  font_size = 8)
+    axs[0].set_title(title1)
+
+    pos = nx.spring_layout(tree2)
+    nx.draw(tree2, pos, with_labels=True, ax=axs[1], node_size=200, node_color='lightgreen', font_size = 8)
+    axs[1].set_title(title2)
+
+    if path:
+        plt.savefig(path)
+    plt.show()
+
+def reconstruct_graph(df, csgraph, root_id):
+    '''
+    This function is used to prune the csgraph created by the skeleton
+    '''
+    valid_ids = set(df['skeleton_id'])
```

Comment:
This is effectively the same algorithm as the one pruning MSTs, but the implementation is totally different. Rewriting the algorithm every time we need to remove nodes from a tree can introduce bugs, so let's try and write a standard algorithm for this (for example, call it prune_nodes_from_tree(graph, nodes_to_remove)) and include it as a static method in the Skeleton class. Then rewrite the pruning functions here to use that common algorithm.